bson/mgocompat/bson_test.go

```diff
@@ -1019,10 +1020,26 @@ func TestUnmarshalSetterErrors(t *testing.T) {
 	assert.Equal(t, "1", m["abc"].Received, "expected m[\"abc\"].Received to be: %v, got: %v", "1", m["abc"].Received)
 }
 
-func TestDMap(t *testing.T) {
-	d := bson.D{{"a", 1}, {"b", 2}}
-	want := bson.M{"a": 1, "b": 2}
-	assert.True(t, reflect.DeepEqual(want, d.Map()), "expected: %v, got: %v", want, d.Map())
+func TestDStringer(t *testing.T) {
+	got := bson.D{{"a", 1}, {"b", 2}}.String()
+	want := `{"a":{"$numberInt":"1"},"b":{"$numberInt":"2"}}`
+	assert.True(t, reflect.DeepEqual(want, got), "expected: %v, got: %v", want, got)
+}
+
+func TestMStringer(t *testing.T) {
+	got := bson.M{"a": 1, "b": 2}.String()
+	m := regexp.MustCompile(`^{(.*),(.*)}$`).FindStringSubmatch(got)
+	assert.Equal(t, 3, len(m), "failed to match %s", got)
+	assert.Equal(t, got, m[0], "expected to match %s, got: %s", got, m[0])
+	wants := map[string]bool{
+		`"a":{"$numberInt":"1"}`: true,
+		`"b":{"$numberInt":"2"}`: true,
+	}
```

Comment:
This assertion logic could be simplified by unmarshaling the resulting string as JSON, then comparing the field values individually. That should be safe because:
* All valid Extended JSON is also valid JSON, so unmarshaling as JSON should always work.
* The "encoding/json" library is more heavily tested than the Extended JSON library and the implementations are independent, so the chance of masking a bug in the function under test should be remote.

For example:
```go
type msg struct {
	A json.RawMessage `json:"a"`
	B json.RawMessage `json:"b"`
}
var res msg
err := json.Unmarshal([]byte(got), &res)
require.NoError(t, err, "Unmarshal error")

want := msg{
	A: json.RawMessage(`{"$numberInt":"1"}`),
	B: json.RawMessage(`{"$numberInt":"2"}`),
}
assert.Equal(t, want, got, "returned string did not unmarshal to the expected document")
```